### PR TITLE
Migrate things from in house codebase

### PR DESCRIPTION
**Pull request by yusup-airwallex**
1.  [x] Upgrade to the in-house implementation
2.  [x] Add static code analyzers 
  - SpotBugs (Upgraded Gradle to version 7)
  - PMD 
3.  [x] Add Github CI build
 - Only push to SonaType from master branch

TODO:
- There are a few things to optimize as suggested by PMD & SpotBugs, will fix those issues in the upcoming PRs.

### CODE REVIEW

File: core/src/test/java/com/airwallex/airskiff/common/GcsDailyLocationTest.java [hunk 1]
```diff
@@ -1,18 +1,39 @@
...
 import java.util.Collections;
 import java.util.List;
-import org.junit.jupiter.api.Test;
-import org.mockito.Mockito;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
 
 public class GcsDailyLocationTest {
 
+  @Disabled
```
Comment (tzhu-airwallex) [inline core/src/test/java/com/airwallex/airskiff/common/GcsDailyLocationTest.java:23]: let's remove this test

File: core/src/test/java/com/airwallex/airskiff/common/GcsDailyLocationTest.java [hunk 2]
```diff
@@ -29,7 +50,7 @@ public void testLatestPath() {
     Mockito.when(blob1.getName()).thenReturn(prefix + "/2020-01-31");
     Mockito.when(blob2.getName()).thenReturn(prefix + "/2020-02-01");
 
-    GcsDailyLocation location = new GcsDailyLocation(storage, bucketName, prefix);
+    com.airwallex.airskiff.common.GcsDailyLocation location = new com.airwallex.airskiff.common.GcsDailyLocation(storage, bucketName, prefix);
```
Comment (tzhu-airwallex) [inline core/src/test/java/com/airwallex/airskiff/common/GcsDailyLocationTest.java:53]: why need full path here?

File: README.md [hunk 3]
```diff
@@ -26,12 +26,12 @@ The library is available in Maven central repository.
 <dependency>
     <groupId>com.airwallex.airskiff</groupId>
     <artifactId>core</artifactId>
-    <version>0.0.2</version>
+    <version>0.0.3</version>
```
Comment (tzhu-airwallex) [inline README.md:29]:
```suggestion
    <version>0.0.4</version>
```

File: README.md [hunk 4]
```diff
@@ -26,12 +26,12 @@ The library is available in Maven central repository.
...
 </dependency>
 ```
 ### Gradle (Kotlin)
 ```gradle
-implementation("com.airwallex.airskiff:core:0.0.2")
+implementation("com.airwallex.airskiff:core:0.0.3")
```
Comment (tzhu-airwallex) [inline README.md:34]:
```suggestion
implementation("com.airwallex.airskiff:core:0.0.4")
```

File: core/src/main/java/com/airwallex/airskiff/common/functions/NamedSerializableLambda.java [hunk 5]
```diff
@@ -1,6 +1,6 @@
 package com.airwallex.airskiff.common.functions;
 
-public interface NamedSerializableLambda<I, O> extends SerializableLambda<I, O> {
+public interface NamedSerializableLambda<I, O> extends com.airwallex.airskiff.common.functions.SerializableLambda<I, O> {
```
[the remark this answers is not in the document]
Comment (yusup-airwallex) [reply core/src/main/java/com/airwallex/airskiff/common/functions/NamedSerializableLambda.java]: fixed.

File: core/src/main/java/com/airwallex/airskiff/core/Either.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
+package com.airwallex.airskiff.core;
+
+import java.io.Serializable;
+
+/**
+ * A Either type that holds one of the two values. Caller must first use isLeft() to decide whether
+ * it is a Left or a Right. Calling the wrong side would throw exceptions. TODO: check if we want to
+ * use Either from https://www.vavr.io
+ */
+public abstract class Either<L, R> implements Serializable {
```
[the remark this answers is not in the document]
Comment (yusup-airwallex) [reply core/src/main/java/com/airwallex/airskiff/core/Either.java:10]: This is still in use.

File: core/src/main/java/com/airwallex/airskiff/flink/RealtimeWatermarkGenerator.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+ * <p>For scenario 3, we will inherit all behaviors from scenario 1 and 2. In addition, we will stop
+ * advance watermark in onPeriodicEmit if there is no event in one period. So if we have an event
+ * coming in that is much earlier than the current processing time, as long as it's later than the
+ * period of its previous event, it will still be processed.
+ *
+ * <p>For scenario 4 and 5, we will always emit a watermark upon arrival of an event. The events
+ * will be accumulated for a while until one event's timestamp is larger than the processing time of
+ * its previous event.
+ */
+public class RealtimeWatermarkGenerator<T> implements WatermarkGenerator<T> {
```
[the remark this answers is not in the document]
Comment (yusup-airwallex) [reply core/src/main/java/com/airwallex/airskiff/flink/RealtimeWatermarkGenerator.java]: Done.

File: core/src/test/java/com/airwallex/airskiff/common/GcsDailyLocationTest.java (hunk 1 above)
Comment (yusup-airwallex) [reply core/src/test/java/com/airwallex/airskiff/common/GcsDailyLocationTest.java]: Done & reverted the commit history as well.

File: core/src/test/java/com/airwallex/airskiff/common/GcsDailyLocationTest.java (hunk 2 above)
Comment (yusup-airwallex) [reply core/src/test/java/com/airwallex/airskiff/common/GcsDailyLocationTest.java]: Removed the whole file.

File: README.md (hunk 3 above)
Comment (yusup-airwallex) [reply README.md]: done.

File: README.md (hunk 4 above)
Comment (yusup-airwallex) [reply README.md]: done.

**Review by tzhu-airwallex — APPROVED** [review]

**Pull request merged**